dlt/destinations/impl/synapse/synapse_adapter.py

```diff
@@ -0,0 +1,50 @@
+from typing import Any, Literal, Set, get_args, Final
+
+from dlt.extract import DltResource, resource as make_resource
+from dlt.extract.typing import TTableHintTemplate
+from dlt.extract.hints import TResourceHints
+from dlt.destinations.utils import ensure_resource
+
+TTableIndexType = Literal["heap", "clustered_columnstore_index"]
+"""
+Table [index type](https://learn.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/sql-data-warehouse-tables-index) used when creating the Synapse table.
+This regards indexes specified at the table level, not the column level.
+"""
+TABLE_INDEX_TYPES: Set[TTableIndexType] = set(get_args(TTableIndexType))
+
+TABLE_INDEX_TYPE_HINT: Literal["x-table-index-type"] = "x-table-index-type"
+
+
+def synapse_adapter(data: Any, table_index_type: TTableIndexType = None) -> DltResource:
+    """Prepares data for the Synapse destination by specifying which table index
+    type should be used.
+
+    Args:
+        data (Any): The data to be transformed. It can be raw data or an instance
+            of DltResource. If raw data, the function wraps it into a DltResource
+            object.
+        table_index_type (TTableIndexType, optional): The table index type used when creating
+            the Synapse table.
+
+    Returns:
+        DltResource: A resource with applied Synapse-specific hints.
+
+    Raises:
+        ValueError: If input for `table_index_type` is invalid.
+
+    Examples:
+        >>> data = [{"name": "Anush", "description": "Integrations Hacker"}]
+        >>> synapse_adapter(data, table_index_type="clustered_columnstore_index")
+        [DltResource with hints applied]
+    """
+    resource = ensure_resource(data)
+
+    if table_index_type is not None:
+        if table_index_type not in TABLE_INDEX_TYPES:
+            allowed_types = ", ".join(TABLE_INDEX_TYPES)
+            raise ValueError(
+                f"Table index type {table_index_type} is invalid. Allowed table index"
+                f" types are: {allowed_types}."
+            )
+        resource._hints[TABLE_INDEX_TYPE_HINT] = table_index_type  # type: ignore[typeddict-unknown-key]
```

Comment:
my take here is to do this with resource.apply_hints() but extend apply_hints to accept a new arg which is a dict which then get's merged into _hints. one of the other freelancers was asking about this (or was it you) so there may be some conflicts. @rudolfix wdyt.

Reply:
so this new arg would be something like additional_hints: DictStrAny.

Reply:
I extended `apply_hints()` to handle table hints as you suggested. Also aligned with @Pipboyguy about this on Slack.